source/mir/ion/internal/stage2.d

```diff
@@ -197,6 +192,7 @@ version (X86_Any)
     }
 
     version (LDC) @target("arch=skylake-avx512")
+    static if (__traits(targetHasFeature, "avx512bw"))
```

Comment:
Shouldn't we simply mark this function `version(none)` and remove other changes?

Reply:
I don't think this is the proper fix. The behavior should be that `@target` compiles for a specific architecture with a specific set of features, regardless of the compilation target. This is the behavior seen if `-mattr=-...` is not passed. 

Passing the flag, however, leads to the compilation error I posted above. This is not inherit to the `avx512bw` function, it can be triggered by any of the functions so long as the right `-mattr` is specified on the command line.

Looking at how rust has it implemented, `@target` should make the function be compiled for the given architecture and it should be able to use intrinsics supported by the target. Looking at the source code of ldc2 I've noticed https://github.com/ldc-developers/ldc/blob/master/gen/uda.cpp#L374-L377. If the `-mattr` from the command line overwrites the `@target` from the function it would explain why it would fail to compile, as the only llvm issues I've read that have similar error messages are due to inappropriate target specifications.

Sure enough, removing the `sort` in ldc2 lets the code compile regardless if I put `-mattr=-avx2` or anything else that previously failed. Note that I had to change the `@target` UDA in the `stage2_impl` function to reference the 
CPU extension that is used instead of the architectures, i.e:
```diff
-    version (LDC) @target("arch=broadwell")
+    version (LDC) @target("avx2")
```
but I think this is acceptable as the `if` I've changed can be reverted and the runtime CPU detection can remain.

I can make the appropriate PRs here and at ldc2 in a couple of days, if that's acceptable.

Reply:
Let's first see what the LDC team thinks about the LDC changes. I am not sure what is the proper fix.

Reply:
I've made the ldc PR and revised the changes in this PR accordingly.

Reply:
What is the mechanic? For example, does `@target("avx2")` add `avx` and `sse4.2`?

Reply:
As far as I can tell yes. I've stumbled over https://doc.rust-lang.org/reference/attributes/codegen.html#x86-or-x86_64 which probably applies to ldc2 `@target` as well.

Reply:
That code is critical for production. I need confirmation from LDC devs or a link to a LLVM codebase.

Please backport the PR for older LDC versions (with `__VERSION__`) when we know which release version will get the required feature. We have to preserve the older behavior for the current compilers.

Reply:
x86 target features hierarchy is described at https://llvm.org/doxygen/X86TargetParser_8cpp.html. So, to specifically answer your previous question:
- [avx2](https://llvm.org/doxygen/X86TargetParser_8cpp.html#adfbe529a682e4bb0a963e2d513b35471) enables avx
- [avx](https://llvm.org/doxygen/X86TargetParser_8cpp.html#a5f4642e45dee81e6c8118ab578dac200) enables sse4.2
- sse4.2 -> sse4.1 -> ssse3 -> sse3 -> sse2 -> sse.

The previous archs are also described as a combination of features, for example https://llvm.org/doxygen/X86TargetParser_8cpp.html#a09b67f1241667fcaace45a872b59d4c6

Reply:
There is also the thing that ldc2 doesn't handle `arch=` in `@target` the same way something like clang does. With clang, specifying `target("arch=<cpu>")` will enable the features supported by the cpu, as they appear in the documentation above. ldc2 doesn't do this, it only sets the CPU field in the function: https://github.com/ldc-developers/ldc/blob/908a2e3d816f3e21d82b61d4b0624de92f871feb/gen/uda.cpp#L368-L371. This seems to be enough (i.e. llvm will generate code using instructions available on the target as specified in the attribute) but it fails if one such target feature is explicitly disabled on the command line.

I believe ldc2 should do the same as clang and generate target features according to the cpu passed in `arch=` but this change doesn't seem trivial because the code in clang that handles it is target dependent:
- For x86: https://github.com/llvm/llvm-project/blob/187c035c047616860d31fc399702538ad1f42755/clang/lib/Basic/Targets/X86.cpp#L117-L119
- For aarch64: https://github.com/llvm/llvm-project/blob/187c035c047616860d31fc399702538ad1f42755/clang/lib/Basic/Targets/AArch64.cpp#L1057-L1066 

Since ldc2's handler code is target independent the changes to fully support all targets won't be that small.

Coming back to this PR I still think the changes from `arch=<cpu>` to `<extension>` are needed because calling a function compiled for 1 cpu on another is probably undefined behavior. Limiting the optimizations to a cpu extension such as `avx2` should get all the optimizations that are needed while allowing more portability as any CPU that implements the extension should be able to safely call the function.

Be aware though that I don't much about this topic. I've done my best reading the code and trying to understand what it does but I may have made mistakes. If you are not 100% convinced that what I say is correct do ask someone with more experience since, as you said, you require this code work.

Reply:
I've been mistaken in my previous comment. Even though ldc2 doesn't expand the cpu features they are expanded in llvm at https://llvm.org/doxygen/MCSubtargetInfo_8cpp_source.html#l00175. Since in the llvm source code the cpu features are set before the general target features their value can be overwritten. This is why, when not specifying `-mattr`, `@target("arch=...")` works properly, but when `-mattr` is used it overwrites the cpu features and that generates an error. This is consistent with clang, however, so it's probably the intended behavior.

Reply:
> Please backport the PR for older LDC versions (with `__VERSION__`) when we know which release version will get the required feature. We have to preserve the older behavior for the current compiler versions.

Done. When the ldc version is at least 1.38.0 the code will use the new `@target` attributes.